common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImplTest.kt

```diff
@@ -0,0 +1,690 @@
+            coVerify(exactly = 2) { mockConnection.send(any<String>()) }
+            assertSame(WebSocketState.ACTIVE, webSocketCore.getConnectionState())
+        }
+
+    @Test
+    fun `Given a valid configuration for 2020 1 1 server When connect is invoked Then it returns true and connection state is ACTIVE`() =
+        runTest {
+            setupServer()
+            prepareAuthenticationAnswer("2020.1.1")
+
+            val result = webSocketCore.connect()
+
+            assertTrue(result)
+            // auth
+            coVerify(exactly = 1) { mockConnection.send(any<String>()) }
+            assertSame(WebSocketState.ACTIVE, webSocketCore.getConnectionState())
+        }
+
+    /*
+sendMessage()
+     */
+
+    @Test
+    fun `Given an active connection that responds to a request When sendMessage is invoked Then it returns a socketResponse and removes the active message`() =
+        runTest {
+            setupServer()
+            prepareAuthenticationAnswer()
+            assertTrue(webSocketCore.connect())
+            val request = mapOf("type" to "test")
+            val expectedMessageSent = webSocketMapper.writeValueAsString(request.plus("id" to 2))
+
+            mockResultSuccessForId(2)
+
+            val response = webSocketCore.sendMessage(request)
+
+            coVerify { mockConnection.send(expectedMessageSent) }
+            assertNotNull(response)
+            assertEquals(2, response.id)
+            assertEquals("result", response.type)
+            assertEquals(emptyMap<Long, WebSocketRequest>(), webSocketCore.activeMessages)
+        }
+
+    @Test
+    fun `Given an active connection that does not respond within timeout When sendMessage is invoked Then it returns null`() =
+        runTest {
+            setupServer()
+            prepareAuthenticationAnswer()
+            assertTrue(webSocketCore.connect())
+            val request = mapOf("type" to "test")
+            val expectedMessageSent = webSocketMapper.writeValueAsString(request.plus("id" to 2))
+
+            val response = webSocketCore.sendMessage(request)
+
+            coVerify { mockConnection.send(expectedMessageSent) }
+            assertNull(response)
+            // TODO It should be empty but currently there is a gap in the impl and the message stays
+            // assertEquals(emptyMap<Long, WebSocketRequest>(), webSocketCore.activeMessages)
+        }
+
+    @Test
+    fun `Given a not connected connection When sendMessage is invoked Then it reconnects`() = runTest {
+        setupServer()
+        prepareAuthenticationAnswer()
+
+        assertNull(webSocketCore.getConnectionState())
+
+        webSocketCore.sendMessage(mapOf("type" to "test"))
+
+        assertEquals(WebSocketState.ACTIVE, webSocketCore.getConnectionState())
+    }
+
+    @Test
+    fun `Given a not connectable connection When sendMessage is invoked Then it returns null and connection state remains null`() = runTest {
+        setupServer("not_connectable")
```

Comment:
"not connectable connection" is a bit vague in my opinion

```suggestion
    fun `Given an invalid url When sendMessage is invoked Then it returns null and connection state remains null`() = runTest {
        setupServer("invalid_url")
```